### DISCUSSION TITLE
@Shared can trigger a view update from a thread other than the main thread, which sometimes causes crash

### DISCUSSION THREAD

**Discussion by larryonoff** [pointfreeco/swift-composable-architecture#3182] [General]
### Description

There’s an issue where updating `@Shared` from a thread other than the main thread triggers a warning during debugging and sometimes causes crashes in the release configuration.

> Store state was accessed on a non-main thread. …
> 
> The "Store" class is not thread-safe, and so all interactions with an instance of "Store" (including all of its scopes and derived view stores) must be done on the main thread

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Crashes or warnings do not occur

### Actual behavior

Crashlytics reports occasional crashes. Xcode shows warnings during debugging.

### Steps to reproduce

Refer to the [sample project](https://github.com/user-attachments/files/15890031/shared-other-than-main.zip)

### The Composable Architecture version information

1.11.1

### Destination operating system

iOS 17.5.1

### Xcode version information

Version 15.4 (15F31d)

### Swift Compiler version information

```shell
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @larryonoff, this will be fixed in #3178 when we require that `withLock` be `@MainActor`. I am going to convert this to a discussion for now, so please feel free to continue the discussion over there!

**Reply by larryonoff** [reply]
Thanks! It was quick!

**Comment by larryonoff** [comment]
A new source of the issue has been identified.

A crash occurs in production when the `PersistenceReaderKey` triggers the `didSet` callback of the `subscribe` function from a non-main thread. This problem arises because SwiftUI updates its views from non-main threads.

The following entry was recorded in Crashlytics:

> SwiftUI/ThreadUtils.swift:114: Fatal error: calling into SwiftUI on a non-main thread is not supported

sample from our code
```swift
  public func subscribe(
    initialValue: Purchases?,
    didSet: @escaping (Purchases?) -> Void
  ) -> Shared<Purchases>.Subscription {
    let task = Task {
      for await purchases in self.purchases.purchasesUpdates() {
        didSet(purchases)

       // looks that possible solution is wrapping didSet into MainActor
       // await MainActor.run { didSet(purchases) }
       // or Task { @MainActor in
      }
    }

    return Shared.Subscription {
      task.cancel()
    }
  }
```

Crashlytics stack trace
<img width="1219" alt="Screenshot 2024-06-21 at 17 23 58" src="https://github.com/pointfreeco/swift-composable-architecture/assets/372616/e527a48e-46bd-4ece-86d3-ffe8dbe3ceeb">

**Reply by larryonoff** [reply]
Warnings were also observed regarding SwiftUI being accessed from a non-main thread when using `@Shared(.appStorage)` `NSUserDefaults` was updated from a non-main thread.

**Reply by larryonoff** [reply]
@mbrandonw It seems that the issue described above deserves GitHub issue.

**Reply by mbrandonw** [reply]
Hi @larryonoff, this should be addressed here: https://github.com/pointfreeco/swift-composable-architecture/pull/3206

**Reply by larryonoff** [reply]
@mbrandonw awesome! thanks for the update
